Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp

```diff
@@ -354,8 +360,19 @@ void MediaPlayerPrivateGStreamerMSE::sourceSetup(GstElement* sourceElement)
 
 void MediaPlayerPrivateGStreamerMSE::updateStates()
 {
-    bool shouldBePlaying = !m_isPaused && readyState() >= MediaPlayer::ReadyState::HaveFutureData;
+    bool shouldBePlaying = !m_isPaused && !m_isPlaybackRatePaused && readyState() >= MediaPlayer::ReadyState::HaveFutureData;
     GST_DEBUG_OBJECT(pipeline(), "shouldBePlaying = %s, m_isPipelinePlaying = %s", boolForPrinting(shouldBePlaying), boolForPrinting(m_isPipelinePlaying));
+
+    GstState state, pending;
+    GstStateChangeReturn getStateResult = gst_element_get_state(pipeline(), &state, &pending, 0);
+    if (getStateResult == GST_STATE_CHANGE_ASYNC) {
+        // Changeing pipeline state during async state change will be either rejected by changePipelineState()
+        // or ignored by playbin (overwritten by the async tartet state after transitions completes).
+        // Lets wait until it's done and then apply play/pause if needed
+        GST_DEBUG_OBJECT(pipeline(), "ASYNC state change in progress");
+        return;
+    }
```

Comment:
We don't see this as a problem. Care to explain why you're adding this?

Reply:
When we do a seek, pipeline gets flushed and sink looses its preroll buffer entering ASYNC paused to paused transition until new preroll buffer is pushed.
In such case the states are:
CURRENT: PAUSED
PENDING: PAUSED
TARGET: PLAYING
Once we call updateStates() at the very moment and playbackRate is set to 0 in meantime we are trying to change the state again into paused as shouldBePlaying is now false and m_isPipelinePlaying is still true.
changePlaybackState(PAUSED) is rejected as it is paused already (PAUSE->PAUSE) but m_isPieplinePlaying is now set to false anyway.

After the seek, pipeline is auto resumed from playbin (applying TARGET state PLAYING) but MSE player thinks it is paused and calling HTMLVideo->pause() does't work any more. Also HTMLVideo->play() doesn't reach GSTMediaPlayer as it reports it is playing.

This can be reproduced with the last test case:
video.currentTime = 10; setTimeout(()=>{video.playbackRate=0;},1);https://asurdej-comcast.github.io/tr/tests/mse_seek_zeroRate.html

Reply:
Just reproduced the same issue again with original way: Highbrow app that after a seek, play/pause doesn't work any more.
This case is not related with playback rates but the flow is the same as described above but initial pause from updateStates() is triggered by HTML. It gets rejected as already in PAUSE->PAUSE transition. After a seek the pipeline is restarted automaticaly but MSE still thinks it is paused. Then GSTMediaPlayer() reports it is not paused (as state is PLAYING) so HTMLvideo->play() doesn't work. And also HTMLVideo->pause() doesn't work as MSE has m_isPipelienPlaying still false.

All of this is cause by rejected state change to paused that still sets m_isPipelinePlaying to false while pipeline is playing in fact.

Reply:
> changePlaybackState(PAUSED) is rejected as it is paused already (PAUSE->PAUSE)

What do you mean exactly with the state change being reverted, and how does that happen? Looking in `gstelement.c`, `gst_element_set_state_func()`, it looks like it should work:

CURRENT: PAUSED
PENDING: PAUSED
TARGET: PLAYING
gst_element_set_state_func(sink, GST_STATE_PAUSED)

Here is the code updating TARGET. I've seen no branches that could avoid it:

    /* this is the (new) state we should go to. TARGET is the last state we set on
    * the element. */
    if (state != GST_STATE_TARGET (element)) {
      GST_CAT_DEBUG_OBJECT (GST_CAT_STATES, element,
          "setting target state to %s", gst_element_state_get_name (state));
      GST_STATE_TARGET (element) = state; // <= TARGET IS UPDATED
      /* increment state cookie so that we can track each state change. We only do
      * this if this is actually a new state change. */
      element->state_cookie++;
    }

Then we would encounter this:

    /* if the element was busy doing a state change, we just update the
    * target state, it'll get to it async then. */
    if (old_pending != GST_STATE_VOID_PENDING) {
      /* upwards state change will happen ASYNC */
      if (old_pending <= state)
        goto was_busy; // <= we get HERE (old_pending=PAUSED, state=PAUSED)
      /* element is going to this state already */
      else if (next == state)
        goto was_busy;
      /* element was performing an ASYNC upward state change and
      * we request to go downward again. Start from the next pending
      * state then. */
      else if (next > state
          && GST_STATE_RETURN (element) == GST_STATE_CHANGE_ASYNC) {
        current = next;
      }
    }

And finally:

    was_busy:
      {
        GST_STATE_RETURN (element) = GST_STATE_CHANGE_ASYNC;
        GST_CAT_DEBUG_OBJECT (GST_CAT_STATES, element,
            "element was busy with async state change");
        GST_OBJECT_UNLOCK (element);

        GST_STATE_UNLOCK (element);

        return GST_STATE_CHANGE_ASYNC;
      }

`GST_STATE_CHANGE_ASYNC` is returned, as we're still in an async state change, we just have changed the goal state.